src/workerd/jsg/filesystem.h

```diff
@@ -0,0 +1,244 @@
+// for this jsg::VirtualFileSystem API.
+class Node: public kj::Refcounted {
+ public:
+  using Type = Node_Type;
+
+  Node() = default;
+  KJ_DISALLOW_COPY_AND_MOVE(Node);
+  virtual ~Node() noexcept(false) = default;
+
+  // Metadata about this filesystem node
+  struct Stat final {
+    Type type = Type::FILE;
+
+    // The size of the node in bytes. For directories, this value will
+    // always be 0.
+    size_t size = 0;
+
+    // The last modified time of the node.
+    kj::Date lastModified = kj::UNIX_EPOCH;
+
+    // The creation time of the node.
+    kj::Date created = kj::UNIX_EPOCH;
+
+    // Indicates if the node is writable.
+    bool writable = false;
+  };
+
+  // Attempt to set the last modified time of this node. If the node is
+  // read-only, this will be a non-op.
+  virtual void setLastModified(Lock& js, kj::Date date = kj::UNIX_EPOCH) = 0;
+
+  // Returns the metadata for this node.
+  virtual const Stat stat() const = 0;
+
+  virtual kj::StringPtr jsgGetMemoryName() const = 0;
+  virtual size_t jsgGetMemorySelfSize() const = 0;
+  virtual void jsgGetMemoryInfo(jsg::MemoryTracker& tracker) const = 0;
+};
+
+// A file in the virtual file system. If the file is read-only, then the
+// mutation methods will throw an exception.
+//
+// A file is writable if it owns its contents. In such cases, the memory
+// allocation is tracked by the corresponding isolate memory accounting.
+class File: public Node {
+ public:
+  kj::Own<File> addRef(jsg::Lock& js) {
+    return kj::addRef(*this);
+  }
+
+  // Reads all the contents of the file as a string. The operation will
+  // fail if the contents exceed a 16 MB limit.
+  kj::String readAllText(Lock& js) const;
+
+  // Reads all the contents of the file as a Uint8Array. The operation will
+  // fail if the contents exceed a 16 MB limit.
+  BufferSource readAllBytes(Lock& js) const;
+
+  // Reads data from the file at the given offset into the given buffer.
+  virtual size_t read(Lock& js, size_t offset, kj::ArrayPtr<kj::byte> buffer) const = 0;
+
+  // Replaces the full contents of the file with the given data.
+  // Equivalent to resize(js, data.size()) followed by write(js, 0, data).
+  void writeAll(Lock& js, kj::ArrayPtr<const kj::byte> data) {
+    resize(js, data.size());
+    write(js, 0, data);
+  }
+
+  // Replaces the full contents of the file with the given data.
+  // Equivalent to resize(js, data.size()) followed by write(js, 0, data).
+  void writeAll(Lock& js, kj::StringPtr data) {
+    writeAll(js, data.asBytes());
+  }
+
+  // Writes data to the file at the given offset. Returns the number of bytes
+  // written.
+  virtual size_t write(Lock& js, size_t offset, kj::ArrayPtr<const kj::byte> data) = 0;
+
+  size_t write(Lock& js, size_t offset, kj::StringPtr data) {
+    return write(js, offset, data.asBytes());
+  }
+
+  // Zero out the contents of the file.
+  virtual void zero(Lock& js, kj::Maybe<size_t> offset = kj::none) = 0;
```

Comment:
seems redundant and if not, seems like it can be polyfilled with write()?

Reply:
That would require at least having an allocation holding the zeroes. Having this as a separate api is a lot more efficient for what is possibly a fairly common case.

Reply:
Changed this to `fill(...)` and added some comments. Intend to keep this as it is far more efficient than using write for the same purpose.